### ISSUE TITLE
Nunjucks shortcode returns undefined when value is an int

### ISSUE THREAD

**Issue by KyleMit** [11ty/eleventy#856]
**Describe the bug**

Nunjucks is having trouble reading the value of a shortcode in some particular cases

**Steps To Reproduce**

1. Start new project
2. Add the following shortcode to `.eleventy.js`

   ```js
   eleventyConfig.addShortcode("year", () => new Date().getFullYear());
   ```
3. Create nunjucks file (`.njk`) and use shortcode `{% year %}`
4. Output is `&copy; undefined`

**Expected behavior**

If you perform step three in a liquid template (`.liquid`) which uses the same shortcode syntax, the output is correct (i.e. `&copy; 2020`)

**Screenshots**

Here's the entire test case in one screenshot

![image](https://user-images.githubusercontent.com/4307307/72300070-bf0f4300-3630-11ea-93b2-091655152a25.png)

**Environment:**

- OS and Version: Windows
- Eleventy Version: 0.10.0

**Additional context**

I thought adding some extra parens (for good measure) might help, but this _doesn't work either_:

```js
eleventyConfig.addShortcode("year", () => ((new Date()).getFullYear()));
```

I thought maybe the inline lamda might have been tripping it up, but this _doesn't work either_:

```js
eleventyConfig.addShortcode("year", () => {
    let year = new Date().getFullYear()
    return year
});
```

I thought maybe the datetime operations might be getting funky, but  this _doesn't work either_:

```js
eleventyConfig.addShortcode("year", () => {
    let year = 2020
    return year
});
```

Turns out - nunjucks doesn't like it when an `int` is returned - and wrapping in a string template literal _will work_:

```js
eleventyConfig.addShortcode("year", () => `${new Date().getFullYear()}`);
```

The problem has a straightforward workaround - so it's not mission critical by any stretch - but thought I'd log in case the fix helps other folks

**Comment by KyleMit** [comment]
Drilled into this a little - the issue occurs in the in the eleventy Nunjucks engine when returning the shortcode value on [L128](https://github.com/11ty/eleventy/blob/v0.10.0/src/Engines/Nunjucks.js#L128)

```js
return new NunjucksLib.runtime.SafeString(shortcodeFn(...argArray));
```

Where `SafeString` is implemented in [nunjucks runtime.js like this](https://github.com/mozilla/nunjucks/blob/master/nunjucks/src/runtime.js#L156-L166):

```js
function SafeString(val) {
  if (typeof val !== 'string') { return val;  }

  this.val = val;
  this.length = val.length;
}
```

The calling API here is a little confusing. If the the val is of type `string`, then the actual object itself isn't returned, so to get the return value, you'd have to use the `new` constructor

```js
NunjucksLib.runtime.SafeString("2020")     // undefined
new NunjucksLib.runtime.SafeString("2020") // SafeString {val: "2020", length: 4}
```

But if we ever wanted to capture the returned value from the function, we wouldn't be able to use the `new` keyword:

```js
NunjucksLib.runtime.SafeString(2020)     // 2020
new NunjucksLib.runtime.SafeString(2020) // SafeString {}
```

If we use the function constructor and don't have a string, the value is never set on the object, so it's not part of the returned object, so I'm not sure when the value from `return val;` is ever used by any nunjucks consumer.

The comment on SafeString is

> A SafeString object indicates that the string should not be autoescaped. This happens magically because autoescaping only occurs on primitive string objects.

So it's purpose isn't to safely convert any object into a string and, in fact, we know that it really isn't handling any object that does not have a `typeof` equal to `string`.  So my guess is we could [cast to a string](https://stackoverflow.com/q/5765398/1366033) prior to passing objects into this method.  At some point, the rendering will have to call `.toString()` somewhere in the pipeline.

```diff
- return new NunjucksLib.runtime.SafeString(shortcodeFn(...argArray));
+ return new NunjucksLib.runtime.SafeString(`${shortcodeFn(...argArray)}`);
```

It's probably reasonable to ask callers to do a string conversion before returning a value, but I also can't see a big downside to forcing the returned object into a string. It *may* not be what the user wants when trying to convert a decimal or date into a string using the default fn, but at least it wouldn't fail with an `undefined` and they could make the change if they wanted explicit control over how it's being mapped into a string.

If that change seems reasonable, I could submit a PR

**Comment by KyleMit** [comment]
Here's a gist with the necessary files to recreate the issue
https://gist.github.com/KyleMit/e8e751550f20997b8269b78b0722edd2

**Comment by pdehaan** [comment]
Yikes. Workaround seems to be:

```js
    eleventyConfig.addShortcode("year", () => {
        let year = new Date().getFullYear();
        return year.toString(); // or `return String(year);`
    });
```

And it looks like Eleventy is using the latest version of Nunjucks (v3.2.1).

**Comment by zachleat** [comment]
Confirmed this as still valid on Eleventy 1.x and 2.x

**Comment by zachleat** [comment]
Fixed, will ship with 2.0.0-canary.13! A lovely comprehensive diagnosis @KyleMit !
